src/plugins/usage_collection/server/collector/types.ts

```diff
@@ -0,0 +1,203 @@
+   */
+  esClient: ElasticsearchClient;
+  /**
+   * Request-scoped Saved Objects client
+   * @remark Bear in mind when testing your collector that your user has the same privileges as the Kibana Internal user to ensure the expected data is sent to the remote cluster (more info: {@link CollectorFetchContext})
+   */
+  soClient: SavedObjectsClientContract;
+} & (WithKibanaRequest extends true
+  ? {
+      /**
+       * The KibanaRequest that can be used to scope the requests:
+       * It is provided only when your custom clients need to be scoped. If not available, you should use the Internal Client.
+       * More information about when scoping is needed: {@link CollectorFetchContext}
+       * @remark You should only use this if you implement your collector to deal with both scenarios: when provided and, especially, when not provided. When telemetry payload is sent to the remote service the `kibanaRequest` will not be provided.
+       */
+      kibanaRequest?: KibanaRequest;
+    }
+  : {});
+
+/**
+ * The fetch method has the context of the Collector itself
+ * (this has access to all the properties of the collector like the logger)
+ * and the the first parameter is {@link CollectorFetchContext}.
+ */
+export type CollectorFetchMethod<
+  WithKibanaRequest extends boolean | undefined,
+  TReturn,
+  ExtraOptions extends object = {}
+> = (
+  this: ICollector<TReturn> & ExtraOptions, // Specify the context of `this` for this.log and others to become available
+  context: CollectorFetchContext<WithKibanaRequest>
+) => Promise<TReturn> | TReturn;
+
+export interface ICollectorOptionsFetchExtendedContext<WithKibanaRequest extends boolean> {
+  /**
+   * Set to `true` if your `fetch` method requires the `KibanaRequest` object to be added in its context {@link CollectorFetchContextWithRequest}.
+   * @remark You should fully understand acknowledge that by using the `KibanaRequest` in your collector, you need to ensure it should specially work without it because it won't be provided when building the telemetry payload actually sent to the remote telemetry service.
+   */
+  kibanaRequest?: WithKibanaRequest;
+}
+
+/**
+ * The options to extend the context provided to the `fetch` method.
+ * @remark Only to be used in very rare scenarios when this is really needed.
+ */
+export type CollectorOptionsFetchExtendedContext<
+  WithKibanaRequest extends boolean
+> = ICollectorOptionsFetchExtendedContext<WithKibanaRequest> &
+  (WithKibanaRequest extends true // If enforced to true via Types, the config must be expected
+    ? Required<Pick<ICollectorOptionsFetchExtendedContext<WithKibanaRequest>, 'kibanaRequest'>>
+    : {});
+
+/**
+ * Options to instantiate a collector
+ */
+export type CollectorOptions<
+  TFetchReturn = unknown,
+  WithKibanaRequest extends boolean = boolean,
+  ExtraOptions extends object = {}
+> = {
+  /**
+   * Unique string identifier for the collector
+   */
+  type: string;
+  init?: Function;
```

Comment:
Any way to be more specific than `Function`?

Reply:
I'm actually wondering if we could remove it (I believe this API is not used at all since it was never documented). I'll see if that statement is true :)